src/colmap/sfm/observation_manager.cc

```diff
@@ -0,0 +1,495 @@
+// Copyright (c) 2023, ETH Zurich and UNC Chapel Hill.
+// All rights reserved.
+//
+// Redistribution and use in source and binary forms, with or without
+// modification, are permitted provided that the following conditions are met:
+//
+//     * Redistributions of source code must retain the above copyright
+//       notice, this list of conditions and the following disclaimer.
+//
+//     * Redistributions in binary form must reproduce the above copyright
+//       notice, this list of conditions and the following disclaimer in the
+//       documentation and/or other materials provided with the distribution.
+//
+//     * Neither the name of ETH Zurich and UNC Chapel Hill nor the names of
+//       its contributors may be used to endorse or promote products derived
+//       from this software without specific prior written permission.
+//
+// THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
+// AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+// IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
+// ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDERS OR CONTRIBUTORS BE
+// LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
+// CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
+// SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+// INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
+// CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+// ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+// POSSIBILITY OF SUCH DAMAGE.
+
+#include "colmap/sfm/observation_manager.h"
+
+#include "colmap/estimators/alignment.h"
+#include "colmap/geometry/triangulation.h"
+#include "colmap/scene/camera.h"
+#include "colmap/scene/projection.h"
+#include "colmap/util/logging.h"
+#include "colmap/util/misc.h"
+
+namespace colmap {
+
+bool MergeAndFilterReconstructions(
+    const double max_reproj_error,
+    const Reconstruction& src_reconstruction,
+    const std::shared_ptr<Reconstruction>& tgt_reconstruction) {
+  if (!MergeReconstructions(
+          max_reproj_error, src_reconstruction, tgt_reconstruction)) {
+    return false;
+  }
+  ObservationManager(tgt_reconstruction)
+      .FilterAllPoints3D(max_reproj_error, /*min_tri_angle=*/0);
+  return true;
+}
+
+const int ObservationManager::kNumPoint3DVisibilityPyramidLevels = 6;
+
+ObservationManager::ObservationManager(
+    std::shared_ptr<Reconstruction> reconstruction,
+    std::shared_ptr<const CorrespondenceGraph> correspondence_graph)
+    : reconstruction_(std::move(THROW_CHECK_NOTNULL(reconstruction))),
+      correspondence_graph_(std::move(correspondence_graph)) {
```

Comment:
Also here THROW_CHECK_NOTNULL?

Reply:
`correspondence_graph` is an optional argument since it is not needed to perform filtering only.